File: src/util/chunker.cpp
#include "flux/util/chunker.hpp"
#include <algorithm>

namespace flux {

std::vector<std::pair<size_t, size_t>> ContentChunker::chunk(std::span<const uint8_t> data) {
    std::vector<std::pair<size_t, size_t>> chunks;
    
    if (data.empty()) {
        return chunks;
    }
    
    size_t offset = 0;
    std::vector<uint8_t> window(WINDOW_SIZE);
    
    while (offset < data.size()) {
        size_t chunk_start = offset;
        size_t chunk_size = 0;
        
        // Scan for chunk boundary
        while (offset < data.size() && chunk_size < MAX_CHUNK_SIZE) {
            chunk_size++;
            offset++;
            
            // Only check for boundaries after minimum chunk size
            if (chunk_size >= MIN_CHUNK_SIZE && offset >= WINDOW_SIZE) {
                // Get rolling window
                size_t window_start = offset - WINDOW_SIZE;
                std::span<const uint8_t> window_data(
                    data.data() + window_start, 
                    WINDOW_SIZE
                );
                
                uint64_t hash = rolling_hash(window_data);
                
                if (is_chunk_boundary(hash)) {
                    break;
                }
            }
        }
        
        chunks.emplace_back(chunk_start, chunk_size);
    }
    
    return chunks;
}

uint64_t ContentChunker::rolling_hash(std::span<const uint8_t> window) {
    uint64_t hash = 0;
    
    for (size_t i = 0; i < window.size(); ++i) {
        hash = (hash * POLYNOMIAL) + window[i];
    }
    
    return hash;
}

bool ContentChunker::is_chunk_boundary(uint64_t hash) {
    // Check if lower bits match pattern (creates ~64KB average chunks)
    constexpr uint64_t mask = (1ULL << 16) - 1;  // 16 bits = ~64KB average
    return (hash & mask) == 0;
}

} // namespace flux

File: src/util/chunker.zig
const std = @import("std");

/// Content-defined chunking for efficient storage of large files
pub const ContentChunker = struct {
    const WINDOW_SIZE: usize = 64;
    const MIN_CHUNK_SIZE: usize = 2 * 1024; // 2 KB
    const MAX_CHUNK_SIZE: usize = 8 * 1024 * 1024; // 8 MB
    const POLYNOMIAL: u64 = 0x3DA3358B4DC173;

    /// Chunk represents a slice of data (offset, length)
    pub const Chunk = struct {
        offset: usize,
        length: usize,
    };

    /// Chunker state for incremental chunking
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) ContentChunker {
        return .{ .allocator = allocator };
    }

    /// Chunk data into content-defined chunks
    pub fn chunk(self: *ContentChunker, data: []const u8) ![]Chunk {
        var chunks = std.ArrayList(Chunk).init(self.allocator);
        errdefer chunks.deinit();

        if (data.len == 0) {
            return chunks.toOwnedSlice();
        }

        var offset: usize = 0;

        while (offset < data.len) {
            const chunk_start = offset;
            var chunk_size: usize = 0;

            // Scan for chunk boundary
            while (offset < data.len and chunk_size < MAX_CHUNK_SIZE) {
                chunk_size += 1;
                offset += 1;

                // Only check for boundaries after minimum chunk size
                if (chunk_size >= MIN_CHUNK_SIZE and offset >= WINDOW_SIZE) {
                    // Get rolling window
                    const window_start = offset - WINDOW_SIZE;
                    const window_data = data[window_start..offset];

                    const hash = rollingHash(window_data);

                    if (isChunkBoundary(hash)) {
                        break;
                    }
                }
            }

            try chunks.append(.{
                .offset = chunk_start,
                .length = chunk_size,
            });
        }

        return chunks.toOwnedSlice();
    }

    /// Compute rolling hash of window
    fn rollingHash(window: []const u8) u64 {
        var hash: u64 = 0;

        for (window) |byte| {
            hash = @mulWithOverflow(hash, POLYNOMIAL)[0] +% byte;
        }

        return hash;
    }

    /// Check if hash indicates chunk boundary
    fn isChunkBoundary(hash: u64) bool {
        // Check if lower bits match pattern (creates ~64KB average chunks)
        const mask: u64 = (1 << 16) - 1; // 16 bits = ~64KB average
        return (hash & mask) == 0;
    }
};

test "content chunking" {
    const allocator = std.testing.allocator;

    var chunker = ContentChunker.init(allocator);

    // Create test data
    const data = try allocator.alloc(u8, 1024 * 1024); // 1 MB
    defer allocator.free(data);

    // Fill with some pattern
    for (data, 0..) |*byte, i| {
        byte.* = @intCast(i % 256);
    }

    const chunks = try chunker.chunk(data);
    defer allocator.free(chunks);

    // Should have at least one chunk
    try std.testing.expect(chunks.len > 0);

    // Verify chunks cover all data
    var total_size: usize = 0;
    for (chunks) |c| {
        total_size += c.length;
    }
    try std.testing.expectEqual(data.len, total_size);
}
